internal/nozzle/nozzle.go

```diff
@@ -88,194 +97,214 @@ func (d *Nozzle) Start() error {
-	c.SetIdleTimeout(time.Duration(d.config.IdleTimeoutSeconds) * time.Second)
+	c.SetIdleTimeout(time.Duration(n.config.IdleTimeoutSeconds) * time.Second)
+	// retry settings
+	c.SetMaxRetryCount(5)
+	c.SetMinRetryDelay(500 * time.Millisecond)
+	c.SetMaxRetryDelay(time.Minute)
 
 	return c, nil
 }
 
-func (d *Nozzle) postToDatadog() error {
-	ticker := time.NewTicker(time.Duration(d.config.FlushDurationSeconds) * time.Second)
-	for {
-		select {
-		case <-ticker.C:
-			d.PostMetrics()
-		case err := <-d.errs:
-			d.handleError(err)
-			return err
-		case <-d.stopper:
-			return nil
-		}
-	}
-}
-
 // Stop stops the Nozzle
-func (d *Nozzle) Stop() {
-	d.stopper <- true
+func (n *Nozzle) Stop() {
+	// We only push value to the `stopper` channel if the Nozzle reading from it.
+	// Hence, if the nozzle is running (`run` method)
+	if !n.isStopped {
+		n.stopper <- true
+	}
 }
 
 // PostMetrics posts metrics do to datadog
-func (d *Nozzle) PostMetrics() {
-	d.mapLock.Lock()
-	// deep copy the metrics map to pass to PostMetrics so that we can unlock d.metricsMap while posting
+func (n *Nozzle) postMetrics() {
+	n.mapLock.Lock()
+	// deep copy the metrics map to pass to PostMetrics so that we can unlock n.metricsMap while posting
 	metricsMap := make(metric.MetricsMap)
-	for k, v := range d.metricsMap {
+	for k, v := range n.metricsMap {
 		metricsMap[k] = v
 	}
-	totalMessagesReceived := d.totalMessagesReceived
+	totalMessagesReceived := n.totalMessagesReceived
 	// Reset the map
-	d.metricsMap = make(metric.MetricsMap)
-	d.mapLock.Unlock()
+	n.metricsMap = make(metric.MetricsMap)
+	n.mapLock.Unlock()
 
 	timestamp := time.Now().Unix()
-	for _, client := range d.ddClients {
+	for _, client := range n.ddClients {
 		// Add internal metrics
 		k, v := client.MakeInternalMetric("totalMessagesReceived", totalMessagesReceived, timestamp)
 		metricsMap[k] = v
-		k, v = client.MakeInternalMetric("totalMetricsSent", d.totalMetricsSent, timestamp)
+		k, v = client.MakeInternalMetric("totalMetricsSent", n.totalMetricsSent, timestamp)
 		metricsMap[k] = v
-		k, v = client.MakeInternalMetric("slowConsumerAlert", atomic.LoadUint64(&d.slowConsumerAlert), timestamp)
+		k, v = client.MakeInternalMetric("slowConsumerAlert", atomic.LoadUint64(&n.slowConsumerAlert), timestamp)
 		metricsMap[k] = v
 
 		err := client.PostMetrics(metricsMap)
+		// NOTE: We don't need to have a retry logic since we don't return error on failure.
+		// However, current metrics may be lost.
 		if err != nil {
-			d.log.Errorf("Error posting metrics: %s\n\n", err)
+			n.log.Errorf("Error posting metrics: %s\n\n", err)
 		}
 	}
 
-	d.totalMetricsSent += uint64(len(metricsMap))
-	d.ResetSlowConsumerError()
+	n.totalMetricsSent += uint64(len(metricsMap))
+	n.ResetSlowConsumerError()
 }
 
-func (d *Nozzle) handleError(err error) {
+func (n *Nozzle) handleError(err error) bool {
+	// If error is a retry error, we log it and let the consumer retry.
 	if retryErr, ok := err.(noaaerrors.RetryError); ok {
```

Comment:
Let's initialize this outside of the `if`, rename the `ok` in `retry`, and reuse `retry` at the end of the function